### PR TITLE
Concept DS added but not finished

### PR DESCRIPTION
**Pull request by lorenzoCapelli**
Concepts with dataset:

- CUB
- AwA
- CIFAR100 with superclasses

Added some minors in attacks and dataset parser

N.B. SUN is missing

### CODE REVIEW

File: peepholelib/adv_atk/attacksDS.py [hunk 1]
```diff
@@ -71,18 +71,18 @@ def apply_attacks(self, **kwargs):
...
-                img_shape = img_sample.shape[1:]
-                with torch.no_grad():
-                    _res = atk.model(img_sample.to(atk.model.device))
-                    num_classes = _res.shape[1]
-
-                self._dss[tdsk]['image'] = MMT.empty(shape=torch.Size((n_samples,)+img_shape))
-                self._dss[tdsk]['label'] = MMT.empty(shape=torch.Size((n_samples,)))
-                self._dss[tdsk]['pred'] = MMT.empty(shape=torch.Size((n_samples,)))
-                self._dss[tdsk]['results'] = MMT.empty(shape=torch.Size((n_samples,)))
+                # img_sample = ds._dss[ds_key]['image'][0:1]
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py]: remove commented old version

File: peepholelib/adv_atk/attacksDS.py [hunk 2]
```diff
@@ -107,14 +107,22 @@ def apply_attacks(self, **kwargs):
...
+
+                keylist = list(ds._dss[ds_key].keys())
+
+                keys_to_copy = [k for k in keylist if k not in keys_to_overwrite]
 
                 for di, dt in tqdm(zip(dl_ori, dl_dst), disable=not verbose, total=ceil(n_samples/bs)): 
                     ori_images = di['image'].to(atk.model.device)
-                    labels = di['label'].int().to(atk.model.device)
+
+                    for key in keys_to_copy:
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py:118]: If we are copying keys, we should not hard-code their type to int here. See comment at line 83.

File: peepholelib/adv_atk/attacksDS.py [hunk 3]
```diff
@@ -71,18 +71,18 @@ def apply_attacks(self, **kwargs):
...
+                # img_sample = ds._dss[ds_key]['image'][0:1]
+                # img_shape = img_sample.shape[1:]
+                # with torch.no_grad():
+                #     _res = atk.model(img_sample.to(atk.model.device))
+                #     num_classes = _res.shape[1]
+
+                for key in ds._dss[ds_key].keys():
+                    key_sample = ds._dss[ds_key][key][0:1]
+                    key_shape = key_sample.shape[1:]
+                    self._dss[tdsk][key] = MMT.empty(shape=torch.Size((n_samples,)+key_shape))
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py]: Get the type of the key sample and create the MMT already with the correct type. So it should work even if the key is not `int`.

File: peepholelib/adv_atk/attacksDS.py [hunk 4]
```diff
@@ -107,14 +107,22 @@ def apply_attacks(self, **kwargs):
...
                     ori_images = di['image'].to(atk.model.device)
-                    labels = di['label'].int().to(atk.model.device)
+
+                    for key in keys_to_copy:
+                        dt[key] = di[key].int()
                     
                     atk_images = atk(
                             images = ori_images,
-                            labels = labels
+                            labels = di['label'].int().to(atk.model.device)
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py]: Same problem here. Looks like labels are not being saved as `int` in the original `parsedDataset`.

File: peepholelib/adv_atk/attacksDS.py [hunk 5]
```diff
@@ -125,10 +133,9 @@ def apply_attacks(self, **kwargs):
 
                     dt['image'] = atk_images
                     dt['output'] = y_pred_atk
-                    dt['label'] = labels
                     dt['pred'] = pred_labels_atk
-                    dt['results'] = pred_labels_atk == labels
-                    dt['attack_success'] = torch.logical_and(pred_labels_ori == labels, pred_labels_atk != pred_labels_ori) 
+                    dt['result'] = pred_labels_atk == dt['label'].to(atk.model.device)
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py]: Can you check if `dt['result']` is in the `cpu` or in `atk.model.device`?

File: peepholelib/datasets/AwA.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+# CIFAR from torchvision
+from torchvision import datasets
+
+def onehot_to_index(bits):
+    for i, b in enumerate(bits):
+        if b == 1:
+            return torch.tensor([i])
+    return torch.tensor([0])
+
+from torch.utils.data import Dataset
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: `os`, `PIL`, `Dataset`, and `torch` are aready imported, clean repeated imports

File: peepholelib/datasets/AwA.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
+# general python stuff
+import os
+from collections import defaultdict
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: unused

File: peepholelib/datasets/AwA.py [hunk 8]
```diff
@@ -0,0 +1,392 @@
+# general python stuff
+import os
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: Use `pathlib.Path` instead of `os`, there is no need for two libraries that do the same thing. As per default we use `pathlib`

File: peepholelib/datasets/AwA.py [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+from collections import defaultdict
+from tqdm import tqdm
+from PIL import Image
+from pathlib import Path
+from types import NoneType
+from math import floor, ceil
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: `floor` seem unused

File: peepholelib/datasets/AwA.py [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+import json
+import re
+
+# Our stuff
+from peepholelib.datasets.parsedDataset import ParsedDataset
+from peepholelib.datasets.datasetWrap import DatasetWrap
+from peepholelib.datasets.functional.transforms import vgg16_cifar100
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: This transform should not be here. Seems unused.

File: peepholelib/datasets/AwA.py [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+import torch
+from torch.utils.data import DataLoader, Dataset
+from torch.utils.data import random_split
+from tensordict import PersistentTensorDict
+from tensordict import MemoryMappedTensor as MMT
+
+# CIFAR from torchvision
+from torchvision import datasets
+
+def onehot_to_index(bits):
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: This function seems unused

File: peepholelib/datasets/AwA.py [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+import os
+import numpy as np
+
+from torch.utils.data import Dataset
+from PIL import Image
+import torch
+import os
+
+class CustomDS(Dataset):
+    def __init__(self, path, reference_ds=None , train=True, train_ratio=0.8, seed=42, transform=None):
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: use `**kwargs` instead of hard-coded named arguments. Will facilitate edits in the future.

File: peepholelib/datasets/AwA.py [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+        
+        assert 0.0 < train_ratio < 1.0
+
+        self.path = path
+        self.reference_ds = reference_ds
+        self.is_train = train
+        self.transform = transform
+
+        # ---- Load class names ----
+        classes_file = os.path.join(path, "classes.txt")
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: use `pathlib.Path` here and in the following ones

File: peepholelib/datasets/AwA.py [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+        img_path = os.path.join(path, "JPEGImages")
+
+        for cid, cname in self.id_to_class.items():
+            class_dir = os.path.join(img_path, cname)
+            if not os.path.isdir(class_dir):
+                print(f"WARNING: folder missing → {class_dir}")
+                continue
+
+            for fname in os.listdir(class_dir):
+                if fname.lower().endswith((".jpg", ".jpeg", ".png")):
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: using the `pathlib.Path` you might want to use the method `glob()` to get the images.

File: peepholelib/datasets/AwA.py [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+        n_total = len(all_samples)
+        n_train = int(train_ratio * n_total)
+
+        g = torch.Generator().manual_seed(seed)
+        perm = torch.randperm(n_total, generator=g)
+
+        train_idx = perm[:n_train].tolist()
+        test_idx  = perm[n_train:].tolist()
+
+        if self.is_train:
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: All samples are being loaded, and then only a portion of them are being returned for `train` or `test`. This requires running all computations twice to obtain both sets, and it is error-prone if the user enters an incorrect seed.

Create the whole dataset as one thing, and during parsing use the `random_split` to create the train and test sets.

File: peepholelib/datasets/AwA.py [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+        return
+
+    def __load_data__(self, **kwargs):
+        """
+        Load and prepare CUB data.
+        """
+        self.__dataset__ = {}
+
+        # Load train split
+        train_dataset = CustomDS(
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: Here, use `random_split()` to divide between test and train

File: peepholelib/datasets/AwA.py [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+            transform=self.transform,
+            reference_ds=self.reference_ds
+        )
+
+        self.__dataset__ = {
+            "train": train_dataset,
+            "test": test_dataset
+        }
+
+        # Build class names dictionary
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: remove unused comments

File: peepholelib/datasets/AwA.py [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+        
+        Returns:
+        - a tuple of (image, label)
+        '''
+        if not self.__dataset__:
+            raise RuntimeError('Data not loaded. Please run load_data() first.')
+        
+        return [self.__dataset__[ds_key][idx]]
+    
+class AwA(ParsedDataset):
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: It feels like this class is unnecessary, and that this dataset could be parsed with `datasets/parsedDataset.parse_ds()`

File: peepholelib/datasets/CUB.py [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+            for line in f:
+                img_id, part_id, x, y, visible = line.strip().split()
+                img_id = int(img_id)
+                part_id = int(part_id)
+                part_info = {
+                    "part_id": part_id,
+                    "part_name": self.part_id_to_name.get(part_id),
+                    "x": float(x),
+                    "y": float(y),
+                    "visible": bool(int(visible)),
```
Comment (lsrosa) [inline peepholelib/datasets/CUB.py]: is that `bool(int())` necessary?

File: peepholelib/datasets/functional/parsers.py [hunk 20]
```diff
@@ -1,10 +1,15 @@
 import torch
+from torch.utils.data._utils.collate import default_collate
```
Comment (lsrosa) [inline peepholelib/datasets/functional/parsers.py]: unused

File: peepholelib/datasets/CUB.py [hunk 21]
```diff
@@ -0,0 +1,442 @@
+# general python stuff
```
Comment (lsrosa) [inline peepholelib/datasets/CUB.py:1]: Apply the same comments from AwA to CUB

**Review by lsrosa — CHANGES_REQUESTED** [review]

File: peepholelib/adv_atk/attacksDS.py [hunk 22]
```diff
@@ -107,14 +105,22 @@ def apply_attacks(self, **kwargs):
                         shuffle = False,
                         num_workers = n_threads
                         )
+                
+                keys_to_overwrite = ['image', 'output', 'pred', 'result']
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py]: pass this as an argument with default values to these values, just in case

File: peepholelib/adv_atk/attacksDS.py [hunk 23]
```diff
@@ -69,20 +69,18 @@ def apply_attacks(self, **kwargs):
...
-
-                self._dss[tdsk]['image'] = MMT.empty(shape=torch.Size((n_samples,)+img_shape))
-                self._dss[tdsk]['label'] = MMT.empty(shape=torch.Size((n_samples,)))
-                self._dss[tdsk]['pred'] = MMT.empty(shape=torch.Size((n_samples,)))
-                self._dss[tdsk]['results'] = MMT.empty(shape=torch.Size((n_samples,)))
+                self._dss[tdsk] = PersistentTensorDict(filename=file_path, batch_size=[n_samples], mode = 'w')
+
+                for key in ds._dss[ds_key].keys():
+                    key_sample = ds._dss[ds_key][key][0:1]
+                    key_shape = key_sample.shape[1:]
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py:78]: could be `key_shape = ds._dss[ds_key][key][0:1].shape[1:]` since `key_sample` is not used, but that is ok

File: peepholelib/adv_atk/attacksDS.py [hunk 24]
```diff
@@ -107,14 +105,22 @@ def apply_attacks(self, **kwargs):
...
                     ori_images = di['image'].to(atk.model.device)
-                    labels = di['label'].int().to(atk.model.device)
+
+                    for key in keys_to_copy:
+                        dt[key] = di[key]
                     
                     atk_images = atk(
                             images = ori_images,
-                            labels = labels
+                            labels = di['label'].to(atk.model.device)
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py]: should we add a `label_key` argument if we decide to apply the atks on superclasses?

File: peepholelib/datasets/AwA.py [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+# from collections import defaultdict
+from tqdm import tqdm
+from PIL import Image
+from pathlib import Path
+# from types import NoneType
+# import json
+# import re
+
+# Our stuff
+# from peepholelib.datasets.parsedDataset import ParsedDataset
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: clean commented imports

File: peepholelib/datasets/AwA.py [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+    
+class AwA(DatasetWrap):
+
+    def __init__(self, **kwargs): 
+        self.path = kwargs.get('path')
+        self.transform = kwargs.get('transform')
+        self.train_ratio = kwargs.get('train_ratio', 0.8)
+        self.seed = kwargs.get('seed', 42)
+        self.reference_ds = kwargs.get('reference_ds', None)
+        self.train_ratio = kwargs.get('train_ratio', 0.8)
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py]: duplicate `train_ratio`

File: peepholelib/datasets/AwA.py [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+        return sample
+    
+class AwA(DatasetWrap):
+
+    def __init__(self, **kwargs): 
+        self.path = kwargs.get('path')
+        self.transform = kwargs.get('transform')
+        self.train_ratio = kwargs.get('train_ratio', 0.8)
+        self.seed = kwargs.get('seed', 42)
+        self.reference_ds = kwargs.get('reference_ds', None)
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py:170]: I do not get why we have a `reference_ds` here; the class should typically load the ds by itself instead of receiving it as an argument. Feels like I am missing something.

File: peepholelib/datasets/CUB.py [hunk 28]
```diff
@@ -0,0 +1,314 @@
...
+import os
+from collections import defaultdict
+from tqdm import tqdm
+from PIL import Image
+from pathlib import Path
+from types import NoneType
+from math import floor, ceil 
+
+# Our stuff
+# from peepholelib.datasets.parsedDataset import ParsedDataset
```
Comment (lsrosa) [inline peepholelib/datasets/CUB.py]: clear commented imports

File: peepholelib/datasets/AwA.py [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+        _ds = CustomDS(
+            path=self.path,
+            transform=self.transform,
+            reference_ds=self.reference_ds,
+            seed=self.seed
+        )
+
+        self.__dataset__ = {}
+        
+        # split train into train and test
```
Comment (lsrosa) [inline peepholelib/datasets/AwA.py:191]: maybe we should split into `train, val, test` since we often need a `val`

File: peepholelib/datasets/CUB.py [hunk 30]
```diff
@@ -0,0 +1,314 @@
...
+            path=self.path,
+            transform=self.transform,
+            reference_ds=self.reference_ds,
+            seed=self.seed
+        )
+
+        self.__dataset__ = {}
+        
+        # split train into train and test
+        self.__dataset__['train'], self.__dataset__['test'] = torch.utils.data.random_split(
```
Comment (lsrosa) [inline peepholelib/datasets/CUB.py]: maybe `train, val, test`

File: peepholelib/datasets/functional/parsers.py [hunk 31]
```diff
@@ -1,10 +1,15 @@
...
-    images, labels = torch.stack(images), torch.tensor(labels)
-    
-    return {'image': images, 'label': labels}
+    features = list(zip(*batch))
 
+    parsed_batch = {}
+
+    for f, key in zip(features, keylist):
+
+        print(f, key)
```
Comment (lsrosa) [inline peepholelib/datasets/functional/parsers.py]: clean print

File: peepholelib/peepholes/classifiers/tgmm.py [hunk 32]
```diff
@@ -92,6 +92,8 @@ def save(self, **kwargs):
         return
 
     def load(self, **kwargs):
+        verbose = kwargs['verbose'] if 'verbose' in kwargs else False
```
Comment (lsrosa) [inline peepholelib/peepholes/classifiers/tgmm.py:92]: `verbose = kwargs.get('verbose', False)`

**Review by lsrosa — CHANGES_REQUESTED** [review]
Very good, just some minor cleanups

File: peepholelib/adv_atk/attacksDS.py (hunk 23 above)
Comment (lorenzoCapelli) [reply peepholelib/adv_atk/attacksDS.py:78]: but we use key_sample is use in line 79 and 80

File: peepholelib/datasets/AwA.py (hunk 27 above)
Comment (lorenzoCapelli) [reply peepholelib/datasets/AwA.py:170]: we have a reference ds in the case of CUB and AwA because i didn't fine tune on those dataset but i use the pretrained model on imagenet

**Review by lsrosa — APPROVED** [review]
